x-pack/filebeat/module/cyberarkpas/audit/ingest/pipeline.yml

```diff
@@ -0,0 +1,1072 @@
+
+        # 428 - Retrieve SSH Key
+        "428":
+          - set: destination.address
+            from: cyberarkpas.audit.ca_properties.address
+          - set: destination.user.name
+            from: cyberarkpas.audit.ca_properties.user_name
+          - set: source.user.name
+            from: cyberarkpas.audit.issuer
+          - set: user.name
+            from: cyberarkpas.audit.issuer
+          - set: event.category
+            value: ["iam"]
+          - set: event.type
+            value: ["admin", "access"]
+          - set: event.outcome
+            value: "success"
+          - set: event.reason
+            from: cyberarkpas.audit.reason
+
+      source: >
+        def clone(def val) {
+            return val instanceof List? new ArrayList(val) : val;
+        }
+        def read_field(def map, String name) {
+          if (map == null || !(map instanceof Map)) return null;
+          int pos = name.indexOf(".");
+          return pos == -1? map[name]
+                          : read_field(map[name.substring(0, pos)], name.substring(pos+1));
+        }
+        String msgID = ctx.event?.code;
+        def actions = params.get(msgID);
+        if (actions == null) return;
+        List values = new ArrayList();
+        for (def item : actions) {
+          def val = item.value;
+          if (val == null && (val = read_field(ctx, item.from)) == null || val == "") continue;
+          values.add([
+            "to": item.set,
+            "value": clone(val)
+          ]);
+        }
+        if (!values.isEmpty()) ctx._tmp["values"] = values;
+
+  - foreach:
+      field: _tmp.values
+      ignore_missing: true
+      processor:
+        set:
+          field: '{{{_ingest._value.to}}}'
+          copy_from: '_ingest._value.value'
+          ignore_empty_value: true
+          override: true
+
+  #
+  # Force event.outcome: unknown in case it gets a value other than one of the allowed.
+  #
+  - set:
+      field: event.outcome
+      value: 'unknown'
+      if: 'ctx.event?.outcome != null && !["success", "failure"].contains(ctx.event.outcome)'
+
+
+  #
+  # Set event.duration from the session duration ("hh:mm:ss") present in some messages.
+  #
+  - script:
+      lang: painless
+      if: "ctx._tmp?.duration_hms != null"
+      source: >
+        long parse_hms(String s) {
+            long cur = 0, total = 0;
+            for (char c: s.toCharArray()) {
+                if (c >= (char)'0' && c <= (char)'9') {
+                    cur = (cur*10) + (long)c - (char)'0';
+                } else if (c == (char)':') {
+                    total = (total + cur) * 60;
+                    cur = 0;
+                } else {
+                    return 0;
+                }
+            }
+            return total + cur;
+        }
+        long nanos = parse_hms(ctx._tmp.duration_hms) * 1000000000L;
+        ctx.event['duration'] = nanos;
+
+  #
+  # Populate ip/domain fields from address.
+  #
+  - grok:
```

Comment:
fyi: 7.13 has an `ip` type in the `convert` processor. https://www.elastic.co/guide/en/elasticsearch/reference/7.x/convert-processor.html

Reply:
👍 using this new method